examples/reinforcement_learning/ddpg.py

```diff
@@ -0,0 +1,396 @@
+
+        valid_rewards /= num_sessions
+        runner.epoch_metrics["_epoch_"]["num_samples"] = self.session_steps
+        runner.epoch_metrics["_epoch_"]["updates_per_sample"] = (
+            runner.loader_sample_step / self.session_steps
+        )
+        runner.epoch_metrics["_epoch_"]["v_reward"] = valid_rewards
+
+
+class CustomRunner(dl.Runner):
+    def __init__(
+        self, *, gamma: float, tau: float, tau_period: int = 1, **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.gamma = gamma
+        self.tau = tau
+        self.tau_period = tau_period
+
+    def on_stage_start(self, runner: dl.IRunner):
+        super().on_stage_start(runner)
+        soft_update(self.model["target_actor"], self.model["actor"], 1.0)
+        soft_update(self.model["target_critic"], self.model["critic"], 1.0)
+
+    def handle_batch(self, batch: Sequence[torch.Tensor]):
+        # model train/valid step
+        states, actions, rewards, dones, next_states = batch
+        actor, target_actor = self.model["actor"], self.model["target_actor"]
+        critic, target_critic = self.model["critic"], self.model["target_critic"]
+        actor_optimizer, critic_optimizer = self.optimizer["actor"], self.optimizer["critic"]
+
+        # get actions for the current state
+        pred_actions = actor(states)
+        # get q-values for the actions in current states
+        pred_critic_states = torch.cat([states, pred_actions], 1)
+        # use q-values to train the actor model
+        policy_loss = (-critic(pred_critic_states)).mean()
+
+        with torch.no_grad():
+            # get possible actions for the next states
+            next_state_actions = target_actor(next_states)
+            # get possible q-values for the next actions
+            next_critic_states = torch.cat([next_states, next_state_actions], 1)
+            next_state_values = target_critic(next_critic_states).detach().squeeze()
+            next_state_values[dones] = 0.0
+
+        # compute Bellman's equation value
+        target_state_values = next_state_values * self.gamma + rewards
+        # compute predicted values
+        critic_states = torch.cat([states, actions], 1)
+        state_values = critic(critic_states).squeeze()
+
+        # train the critic model
+        value_loss = self.criterion(state_values, target_state_values.detach())
+
+        self.batch_metrics.update({"critic_loss": value_loss, "actor_loss": policy_loss})
+
+        if self.is_train_loader:
+            actor.zero_grad()
+            actor_optimizer.zero_grad()
+            policy_loss.backward()
+            actor_optimizer.step()
+
+            critic.zero_grad()
+            critic_optimizer.zero_grad()
+            value_loss.backward()
+            critic_optimizer.step()
+
+            if self.global_batch_step % self.tau_period == 0:
+                soft_update(target_actor, actor, self.tau)
+                soft_update(target_critic, critic, self.tau)
```

Comment:
Is it possible to replace this with an optimizer callback?

Reply:
yes, but for education purposes it's better to have everything in one place